Fix qvm.prefs, qvm.remove

maltesl opened the pull request:
`qubesctl qvm.prefs VMNAME` currently fails due to two old arguments that do not exist anymore in the Qubes core stack (`internal` and `type`). So remove them. Also remove commented-out arguments that were already removed back in R3.0 while we are at it.

maltesl committed:
qvm.prefs: drop deprecated 'internal' and 'type'

These two arguments do not exist anymore in the Qubes core stack, so
running `qubesctl qvm.prefs VMNAME` led to an error. Removing them in
the salt module fixes this.

maltesl committed:
qvm.prefs: remove old comment from Qubes R3.0

maltesl committed:
qvm.remove: fix hanging due to missing --force

A previous version of the command line tool qvm-remove, which is called
by salt through qvm.remove, did not prompt for confirmation.  The
current version does, so qvm.remove hangs. Hence add `--force` to the
call to qvm-remove.

maltesl commented:
I now also added a fix for `qvm.prefs`: Previously it hung because the `qvm-remove` tool that gets run used to not prompt for confirmation, but now it does, and hence requires the `--force` argument for the old behavior.

maltesl commented:
By the way, in the future, for multiple small, but not directly related, changes, do you prefer multiple pull requests (more clutter) or only one (perhaps inconvenient in case one of the parts gets rejected)?

marmarek reviewed (dismissed):
There is `klass` read-only attribute in place of old `type`.

marmarek commented:
Multple small changes in one PR is fine, but please keep them in separate commits (as you've done here).

maltesl committed:
qvm.prefs: add read-only attribute klass

The Qubes core stack replaced the attribute `type` a while ago with the
(read-only) attribute `klass`.
Add `klass` to qvm.prefs as a replacement for `type`, which was removed
from qvm.prefs in d2dd5fb793f2.

maltesl commented:
If I understand correctly the requested change is to add support for `klass`, right? It would also be possible to support `type` here again, and translate it to `klass` as used in the Qubes core stack, but as `type` was broken for a while already (and as far as I know no-one complained), I think just introducing `klass` is better. I added a commit doing that (but I could also add `type` again if you prefer).

I wasn't totally sure what to do if setting the value of `klass` is requested, in the end I decided to `raise SaltInvocationError`, before anything (else) is set.

While testing this I noticed that the way `qvm.prefs` is currently implemented, `klass` will be useless however; It seems to always just return `*default*`.

maltesl commented:
By the way, is there a (technical?) reason for why qubes-mgmt-salt-dom0-qvm re-implements the code for parsing arguments etc., rather than somehow reusing the code from the tools in qubes-core-admin-client? There seems to be a lot of overlap, and if qubes-mgmt-salt-dom0-qvm used things in qubes-core-admin-client directly it wouldn't be necessary to make changes here every time there is a new option.
[Of course, doing this would require at least some changes in both projects, but if time necessary to refactor things appropriately is the only reason for the duplication I might look into it if I have some time (but perhaps I am also missing a good reason for why it is not worth it).]

marmarek commented:
> While testing this I noticed that the way `qvm.prefs` is currently implemented, `klass` will be useless however; It seems to always just return `*default*`.

Oh, indeed, so it's not very useful...

As for overlap with qubes-core-admin-client, there is no technical reason for that, mostly historical. In R3.x days, there was no qubes-core-admin-client providing this kind of API. If you'd like to de-duplicate it, you're very much welcome to do that! But at the same time, try to keep compatibility with existing salt formulas (there are a few of them already in use and we'd like to not force users to change salt files at every qubes update). This may mean introducing few virtual entities (properties, flags etc) translated into something else internally.

marmarek commented:
I've merged it skipping the last commit, as indeed "klass" is not very useful as it is.
